[SPARK-12533][SQL] hiveContext.table() throws the wrong exception

thomastechs opened the pull request:
Avoiding the the No such table exception and throwing analysis exception as per the bug: SPARK-12533

thomastechs committed:
Update Catalog.scala 

Avoiding the the No such table exception and throwing analysis exception as per the bug: SPARK-12533

AmplabJenkins commented:
Can one of the admins verify this patch?

marmbrus commented:
OK to test

rxin commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Catalog.scala:
```diff
@@ -104,7 +104,7 @@ class SimpleCatalog(val conf: CatalystConf) extends Catalog {
     val table = tables.get(tableName)
     if (table == null) {
-      throw new NoSuchTableException
+      throw new AnalysisException("The table named "+tableName+" was not found.")
```
this will fail style checker. you will need to add a space before and after +.

thomastechs replied:
Sure @rxin  .Will update the fix and create a new pull request

thomastechs committed:
Update Catalog.scala- [SPARK-12533]

Incorporating the review comments

thomastechs commented:
@rxin  I have updated the style check difference ,you mentioned. and committed to my branch. And I see it is reflected here in this same pull request. Please review.

SparkQA commented:
**[Test build #2281 has started](https://amplab.cs.berkeley.edu/jenkins/job/NewSparkPullRequestBuilder/2281/consoleFull)** for PR 10529 at commit [`4463a95`](https://github.com/apache/spark/commit/4463a950e6f2b857314cdfd1871449a92ea652af).

SparkQA commented:
**[Test build #2281 has finished](https://amplab.cs.berkeley.edu/jenkins/job/NewSparkPullRequestBuilder/2281/consoleFull)** for PR 10529 at commit [`4463a95`](https://github.com/apache/spark/commit/4463a950e6f2b857314cdfd1871449a92ea652af).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

thomastechs commented:
working on the updates

jayadevanmurali commented:
I think, there would be a reference update in Analyzer.scala as we discussed @thomastechs

thomastechs committed:
Updated Catalog.scala with standard message

AnalysisException is updated as per the bug fic mentions SPARK-12533

thomastechs commented:
@jayadevanmurali : Yes, modifying the  files which has reference to the code fix, as well.

thomastechs committed:
Updated Analyzer.scala with the AnalysisException

As the exception type returned from the lookupRelation method is changed to AnalysisException, it is updated here also.

thomastechs committed:
[SPARK-12438][SQL]Updating CachedTableSuite.scala

Updating the test case methods corresponding the bug fix [SPARK-12438]

thomastechs commented:
@rxin  and @marmbrus and @SparkQA . The fix including the testcases have been committed. Could you please initiate a test build?

SparkQA commented:
**[Test build #2288 has started](https://amplab.cs.berkeley.edu/jenkins/job/NewSparkPullRequestBuilder/2288/consoleFull)** for PR 10529 at commit [`76a9623`](https://github.com/apache/spark/commit/76a9623fed32eeb3c4cd34e6a7ad54ffefca1b69).

SparkQA commented:
**[Test build #2288 has finished](https://amplab.cs.berkeley.edu/jenkins/job/NewSparkPullRequestBuilder/2288/consoleFull)** for PR 10529 at commit [`76a9623`](https://github.com/apache/spark/commit/76a9623fed32eeb3c4cd34e6a7ad54ffefca1b69).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

thomastechs committed:
[SPARK-12533]Updated CachedTableSuite.scala

Removed the AnalysisException class import command, as it throws (hidden) warning. The Analysis Exception is already available in this package

thomastechs committed:
[SPARK-12533]Updated Analyzer.scala

Reverting to change to previous state, because the relevant test cases are to be triggered.

thomastechs commented:
@SparkQA , @rxin  The changes are done to fix the unit test failure. Could you please initiate a test build?

SparkQA commented:
**[Test build #2299 has started](https://amplab.cs.berkeley.edu/jenkins/job/NewSparkPullRequestBuilder/2299/consoleFull)** for PR 10529 at commit [`f75c44e`](https://github.com/apache/spark/commit/f75c44e132bea250ac3c35e3a69bc2fa16d49eb1).

SparkQA commented:
**[Test build #2299 has finished](https://amplab.cs.berkeley.edu/jenkins/job/NewSparkPullRequestBuilder/2299/consoleFull)** for PR 10529 at commit [`f75c44e`](https://github.com/apache/spark/commit/f75c44e132bea250ac3c35e3a69bc2fa16d49eb1).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

rxin commented:
I've merged this. Thanks.

marmbrus commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Catalog.scala:
```diff
@@ -104,7 +104,7 @@ class SimpleCatalog(val conf: CatalystConf) extends Catalog {
     val table = tables.get(tableName)
     if (table == null) {
-      throw new NoSuchTableException
+      throw new AnalysisException("Table not found: " + tableName)
```
Does this mean that `NoSuchTableException` is not used anywhere?  We should probably get rid of it then.  Though, I think I'd have a slight preference for keeping a special exception internally.  The point of this extra control flow is so that we can attach line position information to this specific exception in the analyzer.

dilipbiswal replied:
@marmbrus Shouldn't we have fixed this issue by marking the relation unresolved and let the analyzer generate a proper AnalysisException ? If we were to use the catalog interface for some other purpose like tooling etc  keeping the generic catalog exceptions would have been better ?
